### PR TITLE
win-dll-links: also copy dll from dependencies

### PR DESCRIPTION
**Pull request by brano543**
## Description of changes
Copies Windows dependencies. Refactored changes from https://github.com/NixOS/nixpkgs/pull/227818. DLLs get copied to output path and are correctly symlinked. Tested with wine64 that application can run correctly and link properly.

Fixes https://github.com/NixOS/nixpkgs/issues/38451

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
(I'm not a good person for Windows stuff.)

**Comment by brano543** [comment]
@vcunat I have tested the changes and DLLs get copied to output directory. I have also tested to run with wine64 and executable can run being able to find all dependencies.

**Comment by Artturin** [comment]
`_7zz` used to fail to run, now it runs.

```
tree result/bin
result/bin
├── 7zz.exe
└── mcfgthread-12.dll -> ../../wmgj476qjfw26f9aij1d64lxrjfv6kk0-mcfgthreads-x86_64-w64-mingw32-git/bin/mcfgthread-12.dll
```

File: pkgs/build-support/setup-hooks/win-dll-link.sh [hunk 1]
```diff
@@ -1,45 +1,89 @@
...
+            local listOfDlls=("${!filesToChecks[@]}")
+            local file=${listOfDlls[0]}
+            markFileAsDone "$file"
+            if [ ! -e "./$file" ]; then
+                local pathsFound
+                readarray -d '' pathsFound < <(find "${searchPaths[@]}" -name "$file" -type f -print0)
+                if [ ${#pathsFound[@]} -eq 0 ]; then continue; fi
+                local dllPath
+                dllPath="${pathsFound[0]}"
+                ln -sr "$dllPath" .
```
Comment (Artturin) [inline pkgs/build-support/setup-hooks/win-dll-link.sh:75]:
```suggestion
                CYGWIN+=\ winsymlinks:nativestrict ln -sr "$dllPath" .
```

Comment (Artturin) [reply pkgs/build-support/setup-hooks/win-dll-link.sh:75]: IDK how to use cygwin in nixpkgs, but this was added in https://github.com/nixos/nixpkgs/commit/dd8d0e300064e9d7d0a952c29abb4496800b59c8

**Review by Artturin — APPROVED** [review]
Let's merge quickly

**Pull request merged**